controllers/rollout/canary.go

```diff
@@ -0,0 +1,269 @@
+			return err
+		} else if done {
+			canaryStatus.CurrentStepState = appsv1alpha1.CanaryStepStatePaused
+			klog.Infof("rollout(%s/%s) step(index:%d) state from(%s) -> to(%s)", r.rollout.Namespace, r.rollout.Name,
+				canaryStatus.CurrentStepIndex, appsv1alpha1.CanaryStepStateMetricsAnalysis, canaryStatus.CurrentStepState)
+		}
+
+	case appsv1alpha1.CanaryStepStatePaused:
+		klog.Infof("rollout(%s/%s) run canary strategy, and state(%s)", r.rollout.Namespace, r.rollout.Name, appsv1alpha1.CanaryStepStatePaused)
+		done, err := r.doCanaryPaused()
+		if err != nil {
+			return err
+		} else if done {
+			canaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			canaryStatus.CurrentStepState = appsv1alpha1.CanaryStepStateCompleted
+			klog.Infof("rollout(%s/%s) step(index:%d) state from(%s) -> to(%s)", r.rollout.Namespace, r.rollout.Name,
+				canaryStatus.CurrentStepIndex, appsv1alpha1.CanaryStepStatePaused, canaryStatus.CurrentStepState)
+		}
+
+	case appsv1alpha1.CanaryStepStateCompleted:
+		klog.Infof("rollout(%s/%s) run canary strategy, and state(%s)", r.rollout.Namespace, r.rollout.Name, appsv1alpha1.CanaryStepStateCompleted)
+		if len(r.rollout.Spec.Strategy.Canary.Steps) > int(canaryStatus.CurrentStepIndex) {
+			canaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			canaryStatus.CurrentStepIndex++
+			canaryStatus.CurrentStepState = appsv1alpha1.CanaryStepStateUpgrade
+			klog.Infof("rollout(%s/%s) canary step from(%d) -> to(%d)", r.rollout.Namespace, r.rollout.Name, canaryStatus.CurrentStepIndex-1, canaryStatus.CurrentStepIndex)
+		} else {
+			klog.Infof("rollout(%s/%s) canary run all steps, and completed", r.rollout.Namespace, r.rollout.Name, canaryStatus.CurrentStepIndex-1, canaryStatus.CurrentStepIndex)
+		}
+	}
+
+	return nil
+}
+
+func (r *rolloutContext) doCanaryUpgrade() (bool, error) {
+	steps := len(r.rollout.Spec.Strategy.Canary.Steps)
+	// only traffic routing
+	/*if len(r.rollout.Spec.Strategy.Canary.Steps) == 0 {
+		if r.workload.CanaryReadyReplicas > 0 {
+			klog.Infof("rollout(%s/%s) workload(%s) canaryAvailable(%d), and go to the next stage",
+				r.rollout.Namespace, r.rollout.Name, r.workload.Name, r.workload.CanaryReadyReplicas)
+			return true, nil
+		}
+		klog.Infof("rollout(%s/%s) workload(%s) canaryAvailable(%d), and wait a moment",
+			r.rollout.Namespace, r.rollout.Name, r.workload.Name, r.workload.CanaryReadyReplicas)
+		return false, nil
+	}*/
+	// canary release
+	batchState, err := r.batchControl.BatchReleaseState()
+	if err != nil {
+		return false, err
+	}
+	canaryStatus := r.newStatus.CanaryStatus
+	cond := util.GetRolloutCondition(*r.newStatus, appsv1alpha1.RolloutConditionProgressing)
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and upgrade workload new versions", canaryStatus.CurrentStepIndex, steps)
+	r.newStatus.Message = cond.Message
+	// promote workload next batch release
+	if batchState.Paused || batchState.CurrentBatch < canaryStatus.CurrentStepIndex {
```

Comment:
according to the logic, we will promote the batch if the batchrelease is paused, it sounds against my intution